mqtt/mqtt-bridge/src/persist/memory/mod.rs

```diff
@@ -0,0 +1,192 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use async_trait::async_trait;
+use mqtt3::proto::Publication;
+use parking_lot::Mutex;
+
+use crate::persist::{
+    memory::loader::InMemoryMessageLoader, memory::waking_map::WakingMap, Key, Persist,
+    PersistError,
+};
+
+mod loader;
+mod waking_map;
+
+// In memory persistence implementation used for the bridge
+struct InMemoryPersist {
+    state: Arc<Mutex<WakingMap>>,
+    offset: u32,
+}
+
+#[async_trait]
+impl<'a> Persist<'a> for InMemoryPersist {
+    type Loader = InMemoryMessageLoader;
+
+    fn new() -> Self {
+        let state = WakingMap::new();
+        let state = Arc::new(Mutex::new(state));
+        let offset = 0;
+        InMemoryPersist { state, offset }
+    }
+
+    async fn insert(&mut self, message: Publication) -> Result<Key, PersistError> {
+        let key = Key {
+            offset: self.offset,
+        };
+
+        let mut state_lock = self.state.lock();
+        state_lock.insert(key.clone(), message);
+        self.offset += 1;
+        Ok(key)
+    }
+
+    async fn remove(&mut self, key: Key) -> Result<bool, PersistError> {
+        let mut state_lock = self.state.lock();
+        state_lock
+            .remove_in_flight(&key)
+            .map_err(|_| PersistError::Removal())?;
+        Ok(true)
+    }
+
+    async fn loader(&'a mut self, batch_size: usize) -> InMemoryMessageLoader {
+        InMemoryMessageLoader::new(Arc::clone(&self.state), batch_size).await
+    }
```

Comment:
Discussed with Anca and we think that it is best not to allow multiple separate instances of loader. These will share the same underlying state and will break the waking functionality. I am going to make loader a struct field that get's initialized on InMemoryPersist::new()

Reply:
I now make loader inside the new() and wrap in Arc<Mutex<>>